lusidtools/cocoon/utilities.py

```diff
@@ -1121,5 +1122,155 @@ def create_scope_id(time_generator=None):
     # Multiply by 7 to get value to 100s of nano seconds
     timestamp = hex(int(test * 10000000.0))
     # Create the scope id by joining the hex representation with dashes every 4 characters
-    scope_id = "-".join(timestamp[i : i + 4] for i in range(2, len(timestamp), 4))
+    scope_id = "-".join(timestamp[i: i + 4] for i in range(2, len(timestamp), 4))
     return scope_id
+
+
+def default_fx_forward_model(df: pd.DataFrame, FX_code: str, func_p, func_l,
+                             mapping: dict) -> (pd.DataFrame, dict):
+    """
+    Function that takes 2 rows representing a single forward and merge them into a single transaction
+
+    :param df:  DataFrame containing transactions data
+    :param T_ID:  Column name of transactions IDs
+    :param transaction_type:  Column name of transaction types
+    :param FX_code:  The transaction type that identifies a forward
+    :param func_p:  function that evaluates to true for where the dataframe row is a buy (incoming) side
+    :param func_l: function that evaluates to true for where the dataframe row is a sell (outgoing) side
+    :param mapping_cash_txn: mapping for FX transactions
+    :return:
+    """
+    logging.info(f"combining transactions of type {FX_code} into a single line using {default_fx_forward_model.__name__}"
+                 f" utility function")
+    t_type = mapping["transactions"]["required"]["type"]
+
+    fwds_df = pd.DataFrame(df[df[t_type] == FX_code])
+
+    base_df = fwds_df[func_p]
+    fgn_df = fwds_df[func_l]
+
+    T_ID = mapping["transactions"]["required"]["transaction_id"]
+
+
+    buy_suffix = "_b"
+    sell_suffix = "_s"
+    logging.info(f"merging buy and sell legs of FX trades and suffixing with {[buy_suffix, sell_suffix]}")
+    fwds_txn_df = pd.merge(base_df, fgn_df, how="outer", on=[T_ID, t_type],
+                           suffixes=[buy_suffix, sell_suffix])
+
+    mapping_cash_txn = remap_after_merge(copy.deepcopy(mapping), buy_suffix=buy_suffix, sell_suffix=sell_suffix)
+
+    return fwds_txn_df, mapping_cash_txn
+
+
+def remap_after_merge(mapping: dict, buy_suffix: str, sell_suffix: str, file_type="transactions") -> dict:
```

Comment:
if this is a private function maybe prefix with `__` i.e. `__ remap_after_merge`

Reply:
have left as public and added docstring.

Reply:
also moved copy.deepcopy(mapping) to inside the remapping function so that if users use this as stand along they don't have to worry about accidentally overwriting the original mapping (which is an issue because the merged dataframe is different to the original, hence the original mapping is still required)